R/select_helpers.R

```diff
@@ -20,36 +28,59 @@ NULL
 #' @rdname select_helpers
 #' @export
 all_continuous <- function() {
-  env_variable_type$lst_variable_types %>%
-    purrr::keep(~identical(., "continuous")) %>%
-    names()
+  .generic_selector("variable", "var_type",
+                    .data$var_type %in% "continuous",
+                    fun_name = "all_continuous")
 }
 
 #' @rdname select_helpers
 #' @export
 all_dichotomous <- function() {
-  env_variable_type$lst_variable_types %>%
-    purrr::keep(~identical(., "dichotomous")) %>%
-    names()
+  .generic_selector("variable", "var_type",
+                    .data$var_type %in% "dichotomous",
+                    fun_name = "all_dichotomous")
 }
 
 #' @rdname select_helpers
 #' @export
 all_categorical <- function(dichotomous = TRUE) {
-  types <-
-    switch(dichotomous, c("categorical", "dichotomous")) %||%
-    "categorical"
+  types <- switch(dichotomous, c("categorical", "dichotomous")) %||% "categorical"
 
-  env_variable_type$lst_variable_types %>%
-    purrr::keep(~. %in% types) %>%
-    names()
+  .generic_selector("variable", "var_type",
+                    .data$var_type %in% types,
+                    fun_name = "all_categorical")
 }
 
 #' @rdname select_helpers
 #' @export
 all_interaction <- function() {
-  env_variable_type$lst_variable_types %>%
-    purrr::keep(~identical(., "interaction")) %>%
-    names()
+  .generic_selector("variable", "var_type",
+                    .data$var_type %in% "interaction",
+                    fun_name = "all_interaction")
 }
 
+#' @rdname select_helpers
+#' @export
+all_intercepts <- function() {
+  .generic_selector("variable", "var_type",
+                    .data$var_type %in% "intercept",
+                    fun_name = "all_intercepts")
+}
+
+#' @rdname select_helpers
+#' @export
+all_contrasts <- function(type = c("treatment", "sum", "poly", "helmert")) {
+  type <- match.arg(type, several.ok = TRUE)
+  contr.type <-
+    purrr::map_chr(type,
+                   ~switch(.x,
```

Comment:
For information, it could also be "contr.SAS" or "custom". Funny fact, SAS is a type of treatment contrasts.
And you can have sometimes "contr.treatement(base=3)".

Maybe it could help if I add in `tidy_add_contrats()` a 'contrasts_type' column with just the options "treatment", "sum", "poly", "helmert" or "custom"

Reply:
I just updated tidy_add_contrasts(), cf. https://github.com/larmarange/broom.helpers/commit/f149841fedd95b890e9f96502d776cb312437097

In addition to `contrasts` column who keeps details, it also add a `contrasts_type` column equal to
"treatment", "sum", "poly", "helmert" or "other". The select helpers could use this column as it will be much easier. What do you think?

Reply:
We can remove the `match.args()` from the function so users can pass any contrast they like.  We can pre-populate the argument with the ones we know about, but if there are others they can easily be handled.

```r
#' @rdname select_helpers
#' @export
all_contrasts <- function(type = c("treatment", "sum", "poly", "helmert", "SAS", "custom")) {
  contr.type <-
    purrr::map_chr(type,
                   ~switch(.x,
                           "treatment" = "contr.treatment",
                           "sum" = "contr.sum",
                           "poly" = "contr.poly",
                           "SAS" = "contr.SAS",
                           "helmert" = "contr.helmert",
                           "custom" = "custom") %||%
                     .x
    )
  
  .generic_selector("variable", "contrasts",
                    .data$contrasts %in% contr.type,
                    fun_name = "all_contrasts")
}
```

Reply:
Perhaps it'll be more clear if we remove the `switch()` and the argument values will match the actual contrast values.  Then the function would just be:

```r
#' @rdname select_helpers
#' @export
all_contrasts <- function(type = c("contr.treatment", "contr.sum", "contr.poly", "contr.helmert", "contr.SAS", "contr.treatement(base=3)", "custom")) {
 
  .generic_selector("variable", "contrasts",
                    .data$contrasts %in% type,
                    fun_name = "all_contrasts")
}
```

Reply:
Last (and perhaps my favorite suggestion). We leave `type=NULL` by default, which will select all contrasts (i.e. `!is.na(contrasts)`).  Then allow users to pass any specific contrasts in type if they only want a particular type.

Reply:
I like your idea of proposing a `NULL` option. As I introduced a `contrasts_type` column in `broom.helpers`, I would suggest to rename `type` argument into `contrasts`, i.e. `all_contrasts(contrasts = NULL)`.

Then we could also create a function `all_contrasts_type(type = c("treatment", "sum", "poly", "helmert", "other"))` with a `match.args()`. (It's now very easy with your generic functions). Therefore we would cover all situations.

What do you think?

Reply:
I'll update!

For the `"contrast_type"` column, do you think it would makes sense/be possible to have it populated for all levels of the variable (like the `"contrast"` column is).

``` r
library(broom.helpers)

glm(
  response ~ stage + grade + trt + factor(death),
  gtsummary::trial,
  family = binomial,
  contrasts = list(
    stage = contr.treatment(4, 3), grade = contr.treatment(3, 2),
    trt = contr.treatment(2, 2), "factor(death)" = matrix(c(-3, 2))
  )
) %>%
  tidy_plus_plus(add_header_rows = TRUE) %>%
  dplyr::select(term, variable, label, contrasts, contrasts_type)
#> # A tibble: 14 x 5
#>    term         variable     label             contrasts          contrasts_type
#>    <chr>        <chr>        <chr>             <chr>              <chr>         
#>  1 <NA>         stage        T Stage           contr.treatment(b~ <NA>          
#>  2 stage1       stage        T1                contr.treatment(b~ treatment     
#>  3 stage2       stage        T2                contr.treatment(b~ treatment     
#>  4 stage3       stage        T3                contr.treatment(b~ <NA>          
#>  5 stage4       stage        T4                contr.treatment(b~ treatment     
#>  6 <NA>         grade        Grade             contr.treatment(b~ <NA>          
#>  7 grade1       grade        I                 contr.treatment(b~ treatment     
#>  8 grade2       grade        II                contr.treatment(b~ <NA>          
#>  9 grade3       grade        III               contr.treatment(b~ treatment     
#> 10 <NA>         trt          Chemotherapy Tre~ contr.SAS          <NA>          
#> 11 trt1         trt          Drug A            contr.SAS          treatment     
#> 12 trt2         trt          Drug B            contr.SAS          <NA>          
#> 13 <NA>         factor(deat~ factor(death)     custom             <NA>          
#> 14 factor(deat~ factor(deat~ factor(death)1    custom             other
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
One very small thing....I think the column names sound better when spoken without the S at the end, e.g. `c("contrast", "contrast_type")`.  Does not really matter though!

Reply:
OK! I just pushed an update. Here's what I implemented.
- I made a single `all_contrasts(contrasts_type=)` function.
- The argument defaults to NULL and selects variables with any type of contrast.
- The argument name matches the column name from the output. If a user does specify one or more contrasts, the input is passed through the `match.args()` to ensure the user selects an appropriate types (`c("treatment", "sum", "poly", "helmert", "other")`)

Are you OK with this implementation?  I like that we can achieve both goals in a single function.

Questions remaining for you:
1. What do you think about updating the column names to `c("contrast", "contrast_type")`?
1. For the "contrast_type" column, do you think it would makes sense/be possible to have it populated for all levels of the variable (like the "contrast" column is)?

Reply:
> For the `"contrast_type"` column, do you think it would makes sense/be possible to have it populated for all levels of the variable (like the `"contrast"` column is).

You are right. Should be fixed by https://github.com/larmarange/broom.helpers/commit/51bc325351cf6a2ecba8a8640d47d8881dc34195

``` r
library(broom.helpers)

glm(
  response ~ stage + grade + trt + factor(death),
  gtsummary::trial,
  family = binomial,
  contrasts = list(
    stage = contr.treatment(4, 3), grade = contr.treatment(3, 2),
    trt = contr.treatment(2, 2), "factor(death)" = matrix(c(-3, 2))
  )
) %>%
  tidy_plus_plus(add_header_rows = TRUE) %>%
  dplyr::select(term, variable, label, contrasts, contrasts_type)
#> # A tibble: 14 x 5
#>    term         variable     label             contrasts          contrasts_type
#>    <chr>        <chr>        <chr>             <chr>              <chr>         
#>  1 <NA>         stage        T Stage           contr.treatment(b~ treatment     
#>  2 stage1       stage        T1                contr.treatment(b~ treatment     
#>  3 stage2       stage        T2                contr.treatment(b~ treatment     
#>  4 stage3       stage        T3                contr.treatment(b~ treatment     
#>  5 stage4       stage        T4                contr.treatment(b~ treatment     
#>  6 <NA>         grade        Grade             contr.treatment(b~ treatment     
#>  7 grade1       grade        I                 contr.treatment(b~ treatment     
#>  8 grade2       grade        II                contr.treatment(b~ treatment     
#>  9 grade3       grade        III               contr.treatment(b~ treatment     
#> 10 <NA>         trt          Chemotherapy Tre~ contr.SAS          treatment     
#> 11 trt1         trt          Drug A            contr.SAS          treatment     
#> 12 trt2         trt          Drug B            contr.SAS          treatment     
#> 13 <NA>         factor(deat~ factor(death)     custom             other         
#> 14 factor(deat~ factor(deat~ factor(death)1    custom             other
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
This is a good question about `contrasts_type` vs `contrast_type`.

In all the stat and R literature, I always found contrasts in a plural form. For one variable, you define contrasts (and not a contrast). This is why I always used the plural form.

However, I'm not a native English speaker.

Reply:
> * I made a single `all_contrasts(contrasts_type=)` function.

That's OK for me

Reply:
> This is a good question about `contrasts_type` vs `contrast_type`.
> 
> In all the stat and R literature, I always found contrasts in a plural form. For one variable, you define contrasts (and not a contrast). This is why I always used the plural form.
> 
> However, I'm not a native English speaker.

When I read about them, we typically are dealing with a set of contrasts, but we certainly can have a single contrast. This page is the first when I google regression model contrasts, and uses the singular where appropriate.
https://online.stat.psu.edu/stat555/node/73/#:~:text=A%20contrast%20is%20essentially%20a,general%20hypotheses%20about%20population%20means.

*This is soooo not a big deal,* but I think the singular would be more accurate. Either is totally fine, and making updates to variable names is annoying!  It can stay as is :)

Reply:
Thanks for the clarification.

- For `contrasts` column, it would be a minor break change with version 1.0.0 and will require to review the code properly.
- For `contrasts_type`, it is very easy to change for `contrast_type`.

Would it be strange to have `contrasts` and `contrast_type`?

Reply:
Yeah, I think either keep contrasts and contrasts_type, or change both to contrast and contrast_type.

Reply:
OK. Let keeps current names to avoid side effects.